File: android/app/src/main/java/com/albbiz/map/ui/screens/EventsScreen.kt
// Bismillah Hir Rahman Nir Raheem
package com.albbiz.map.ui.screens

import android.content.Intent
import android.net.Uri
import android.widget.Toast
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.compose.AsyncImage
import com.albbiz.map.data.Event
import com.albbiz.map.data.EventsRepository
import com.albbiz.map.ui.LocalAppStrings
import com.albbiz.map.ui.MeTontGrey
import com.albbiz.map.ui.MeTontRed
import java.text.SimpleDateFormat
import java.util.*

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun EventsScreen(
    onBackClick: () -> Unit,
    onAddEventClick: () -> Unit
) {
    val repository = remember { EventsRepository() }
    // getEvents() must be remember()'d too, not just the repository — calling it
    // fresh inline on every recomposition hands collectAsState() a brand-new Flow
    // object each time, and it treats a new Flow instance as a reason to cancel the
    // old collection and start over. That was tearing down and reattaching a live
    // Firestore listener on every recomposition of this screen instead of once.
    val eventsFlow = remember { repository.getEvents() }
    val events by eventsFlow.collectAsState(initial = emptyList())
    val strings = LocalAppStrings.current

    Scaffold(
        topBar = {
            TopAppBar(
                title = {
                    Text(
                        strings.communityEventsTitle,
                        color = Color.White,
                        fontWeight = FontWeight.Bold
                    )
                },
                navigationIcon = {
                    IconButton(onClick = onBackClick) {
                        Icon(
                            Icons.AutoMirrored.Filled.ArrowBack,
                            "Back",
                            tint = Color.White
                        )
                    }
                },
                colors = TopAppBarDefaults.topAppBarColors(
                    containerColor = MeTontRed
                )
            )
        },
        floatingActionButton = {
            FloatingActionButton(
                onClick = onAddEventClick,
                containerColor = MeTontRed,
                contentColor = Color.White,
                shape = CircleShape
            ) {
                Icon(Icons.Default.Add, "Add Event")
            }
        }
    ) { padding ->
        Box(
            modifier = Modifier
                .fillMaxSize()
                //.background(Color(0xFFFFF8F0))
                .background(MaterialTheme.colorScheme.background)
                .padding(padding)
        ) {
            if (events.isEmpty()) {
                // ── EMPTY STATE ───────────────────────────────────
                Box(
                    modifier = Modifier.fillMaxSize(),
                    contentAlignment = Alignment.Center
                ) {
                    Column(
                        horizontalAlignment = Alignment.CenterHorizontally,
                        verticalArrangement = Arrangement.spacedBy(12.dp)
                    ) {
                        Surface(
                            modifier = Modifier.size(80.dp),
                            shape = CircleShape,
                            color = MeTontRed.copy(alpha = 0.1f)
                        ) {
                            Box(contentAlignment = Alignment.Center) {
                                Icon(
                                    Icons.Default.CalendarMonth,
                                    null,
                                    modifier = Modifier.size(40.dp),
                                    tint = MeTontRed.copy(alpha = 0.5f)
                                )
                            }
                        }
                        Text(
                            strings.noEventsFound,
                            style = MaterialTheme.typography.bodyLarge,
                            color = MeTontGrey,
                            fontWeight = FontWeight.Medium
                        )
                        Text(
                            "Be the first to add a community event!",
                            style = MaterialTheme.typography.bodySmall,
                            color = MeTontGrey.copy(alpha = 0.7f)
                        )
                    }
                }
            } else {
                LazyColumn(
                    modifier = Modifier.fillMaxSize(),
                    contentPadding = PaddingValues(16.dp),
                    verticalArrangement = Arrangement.spacedBy(16.dp)
                ) {
                    item {
                        Text(
                            "${events.size} upcoming ${if (events.size == 1) "event" else "events"}",
                            color = MeTontGrey,
                            fontSize = 13.sp,
                            modifier = Modifier.padding(bottom = 4.dp)
                        )
                    }
                    items(events) { event ->
                        EventItem(event)
                    }
                }
            }
        }
    }
}

@Composable
fun EventItem(event: Event) {
    val strings = LocalAppStrings.current
    val context = LocalContext.current
    // Events don't carry enough extra data (unlike a business's reviews/photos/hours)
    // to justify a whole separate detail screen — the only thing actually missing
    // was a way to read a description longer than 3 lines. Expanding in place on tap
    // covers that without adding a new screen.
    var expanded by remember { mutableStateOf(false) }

    Card(
        modifier = Modifier.fillMaxWidth(),
        shape = RoundedCornerShape(16.dp),
        colors = CardDefaults.cardColors(containerColor = Color.White),
        elevation = CardDefaults.cardElevation(defaultElevation = 2.dp)
    ) {
        Column {
            // ── EVENT IMAGE ───────────────────────────────────────
            if (event.imageUrl != null) {
                AsyncImage(
                    model = event.imageUrl,
                    contentDescription = null,
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(180.dp)
                        .clip(RoundedCornerShape(topStart = 16.dp, topEnd = 16.dp)),
                    contentScale = ContentScale.Crop
                )
            } else {
                // Placeholder if no image
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(100.dp)
                        .background(MeTontRed.copy(alpha = 0.1f)),
                    contentAlignment = Alignment.Center
                ) {
                    Icon(
                        Icons.Default.Event,
                        null,
                        modifier = Modifier.size(48.dp),
                        tint = MeTontRed.copy(alpha = 0.4f)
                    )
                }
            }

            Column(modifier = Modifier.padding(16.dp)) {
                // ── TITLE + PROMOTED BADGE ────────────────────────
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.Top
                ) {
                    Text(
                        event.title,
                        style = MaterialTheme.typography.titleMedium,
                        fontWeight = FontWeight.Bold,
                        color = Color.Black,
                        modifier = Modifier.weight(1f)
                    )
                    if (event.isPromoted) {
                        Spacer(Modifier.width(8.dp))
                        Surface(
                            color = Color(0xFFFFC107),
                            shape = RoundedCornerShape(6.dp)
                        ) {
                            Text(
                                strings.promoted,
                                modifier = Modifier.padding(
                                    horizontal = 8.dp,
                                    vertical = 3.dp
                                ),
                                style = MaterialTheme.typography.labelSmall,
                                fontWeight = FontWeight.Bold,
                                color = Color.White
                            )
                        }
                    }
                }

                // ── CATEGORY CHIP ─────────────────────────────────
                Surface(
                    color = MeTontRed.copy(alpha = 0.1f),
                    shape = RoundedCornerShape(6.dp),
                    modifier = Modifier.padding(top = 4.dp)
                ) {
                    Text(
                        event.category,
                        modifier = Modifier.padding(horizontal = 8.dp, vertical = 3.dp),
                        style = MaterialTheme.typography.labelSmall,
                        color = MeTontRed,
                        fontWeight = FontWeight.Medium
                    )
                }

                Spacer(Modifier.height(12.dp))

                // ── DATE ──────────────────────────────────────────
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Icon(
                        Icons.Default.Event,
                        null,
                        modifier = Modifier.size(16.dp),
                        tint = MeTontRed
                    )
                    Spacer(Modifier.width(6.dp))
                    val date = SimpleDateFormat(
                        "EEEE, MMM dd, yyyy",
                        Locale.getDefault()
                    ).format(Date(event.date))
                    Text(
                        date,
                        style = MaterialTheme.typography.bodySmall,
                        color = MeTontGrey
                    )
                }

                Spacer(Modifier.height(6.dp))

                // ── LOCATION ──────────────────────────────────────
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Icon(
                        Icons.Default.LocationOn,
                        null,
                        modifier = Modifier.size(16.dp),
                        tint = MeTontRed
                    )
                    Spacer(Modifier.width(6.dp))
                    Text(
                        event.locationName,
                        style = MaterialTheme.typography.bodySmall,
                        color = MeTontGrey
                    )
                }

                Spacer(Modifier.height(10.dp))

                // ── DESCRIPTION ───────────────────────────────────
                // Tapping the description toggles between a 3-line preview and the
                // full text, instead of a separate detail screen — see the comment
                // above the expanded state declaration for why.
                Text(
                    event.description,
                    style = MaterialTheme.typography.bodySmall,
                    color = Color.Black.copy(alpha = 0.7f),
                    maxLines = if (expanded) Int.MAX_VALUE else 3,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.clickable { expanded = !expanded }
                )
                Text(
                    if (expanded) strings.readLess else strings.readMore,
                    style = MaterialTheme.typography.labelSmall,
                    color = MeTontRed,
                    fontWeight = FontWeight.Bold,
                    modifier = Modifier
                        .padding(top = 4.dp)
                        .clickable { expanded = !expanded }
                )

                // ── WEBSITE BUTTON ────────────────────────────────
                if (event.websiteUrl != null) {
                    Spacer(Modifier.height(8.dp))
                    OutlinedButton(
                        onClick = {
                            try {
                                val url = event.websiteUrl.let {
                                    if (it.startsWith("http://") || it.startsWith("https://")) it
                                    else "https://$it"
                                }
                                context.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
                            } catch (e: Exception) {
                                Toast.makeText(context, "Couldn't open that link", Toast.LENGTH_SHORT).show()
                            }
                        },
                        shape = RoundedCornerShape(8.dp),
                        colors = ButtonDefaults.outlinedButtonColors(
                            contentColor = MeTontRed
                        ),
                        border = androidx.compose.foundation.BorderStroke(
                            1.dp, MeTontRed
                        ),
                        contentPadding = PaddingValues(
                            horizontal = 12.dp,
                            vertical = 6.dp
                        )
                    ) {
                        Icon(
                            Icons.Default.Language,
                            null,
                            modifier = Modifier.size(14.dp)
                        )
                        Spacer(Modifier.width(4.dp))
                        Text(
                            strings.viewEventWebsite,
                            style = MaterialTheme.typography.labelMedium
                        )
                    }
                }
            }
        }
    }
}

File: ios/AlbBizMap_IOS/AlbBizMap/UI/Screens/EventsScreen.swift
// Bismillah Hir Rahman Nir Raheem
import SwiftUI
import UIKit

public struct EventsScreen: View {
    @Environment(\.appStrings) private var strings

    @ObservedObject private var repo = EventsRepository.shared
    public let onBackClick: () -> Void
    public let onAddEventClick: () -> Void

    public init(onBackClick: @escaping () -> Void, onAddEventClick: @escaping () -> Void) {
        self.onBackClick = onBackClick
        self.onAddEventClick = onAddEventClick
    }

    public var body: some View {
        VStack(spacing: 0) {
            topAppBar

            ZStack(alignment: .bottomTrailing) {
                if repo.events.isEmpty {
                    emptyState
                } else {
                    ScrollView {
                        VStack(alignment: .leading, spacing: 16) {
                            Text("\(repo.events.count) upcoming \(repo.events.count == 1 ? "event" : "events")")
                                .font(.system(size: 13))
                                .foregroundColor(.meTontGrey)

                            ForEach(repo.events) { evt in
                                EventItemCard(event: evt)
                            }
                        }
                        .padding(16)
                    }
                }

                Button(action: onAddEventClick) {
                    Image(systemName: "plus")
                        .font(.system(size: 22, weight: .bold))
                        .foregroundColor(.white)
                        .frame(width: 56, height: 56)
                        .background(Color.meTontRed)
                        .clipShape(Circle())
                        .shadow(color: .black.opacity(0.25), radius: 6, x: 0, y: 3)
                }
                .padding(20)
            }
        }
        .background(Color.meTontBackground)
    }

    private var topAppBar: some View {
        HStack(spacing: 8) {
            Button(action: onBackClick) {
                Image(systemName: "chevron.left").foregroundColor(.white)
            }
            Text(strings.communityEventsTitle)
                .font(.system(size: 18, weight: .bold))
                .foregroundColor(.white)
            Spacer()
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 12)
        .background(Color.meTontRed)
    }

    private var emptyState: some View {
        VStack(spacing: 12) {
            ZStack {
                Circle().fill(Color.meTontRed.opacity(0.1)).frame(width: 80, height: 80)
                Image(systemName: "calendar")
                    .font(.system(size: 34))
                    .foregroundColor(.meTontRed.opacity(0.5))
            }
            Text(strings.noEventsFound)
                .font(.system(size: 16, weight: .medium))
                .foregroundColor(.meTontGrey)
            Text("Be the first to add a community event!")
                .font(.system(size: 13))
                .foregroundColor(.meTontGrey.opacity(0.7))
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

private struct EventItemCard: View {
    @Environment(\.appStrings) private var strings
    let event: Event
    @State private var expanded: Bool = false

    private var formattedDate: String {
        let date = Date(timeIntervalSince1970: TimeInterval(event.date) / 1000.0)
        let formatter = DateFormatter()
        formatter.dateFormat = "EEEE, MMM d, yyyy 'at' h:mm a"
        return formatter.string(from: date)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            // Event image / placeholder
            if let imgUrl = event.imageUrl, let url = URL(string: imgUrl) {
                AsyncImage(url: url) { image in
                    image.resizable().aspectRatio(contentMode: .fill)
                } placeholder: {
                    Color.gray.opacity(0.15)
                }
                .frame(height: 180)
                .clipped()
            } else {
                ZStack {
                    Color.meTontRed.opacity(0.1)
                    Image(systemName: "calendar")
                        .font(.system(size: 40))
                        .foregroundColor(.meTontRed.opacity(0.4))
                }
                .frame(height: 100)
            }

            VStack(alignment: .leading, spacing: 0) {
                // Title + Promoted badge
                HStack(alignment: .top, spacing: 8) {
                    Text(event.title)
                        .font(.system(size: 17, weight: .bold))
                        .foregroundColor(.black)
                        .frame(maxWidth: .infinity, alignment: .leading)

                    if event.isPromoted {
                        Text(strings.promoted)
                            .font(.system(size: 10, weight: .bold))
                            .foregroundColor(.white)
                            .padding(.horizontal, 8)
                            .padding(.vertical, 3)
                            .background(Color(red: 0xFF/255.0, green: 0xC1/255.0, blue: 0x07/255.0))
                            .cornerRadius(6)
                    }
                }

                // Category chip
                Text(event.category)
                    .font(.system(size: 11, weight: .medium))
                    .foregroundColor(.meTontRed)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 3)
                    .background(Color.meTontRed.opacity(0.1))
                    .cornerRadius(6)
                    .padding(.top, 4)

                Spacer().frame(height: 12)

                // Date
                HStack(spacing: 6) {
                    Image(systemName: "calendar")
                        .font(.system(size: 14))
                        .foregroundColor(.meTontRed)
                    Text(formattedDate)
                        .font(.system(size: 13))
                        .foregroundColor(.meTontGrey)
                }

                Spacer().frame(height: 6)

                // Location
                HStack(spacing: 6) {
                    Image(systemName: "mappin.circle.fill")
                        .font(.system(size: 14))
                        .foregroundColor(.meTontRed)
                    Text(event.locationName)
                        .font(.system(size: 13))
                        .foregroundColor(.meTontGrey)
                }

                Spacer().frame(height: 10)

                // Description
                Text(event.description)
                    .font(.system(size: 13))
                    .foregroundColor(.black.opacity(0.7))
                    .lineLimit(expanded ? nil : 3)
                    .onTapGesture { expanded.toggle() }

                Text(expanded ? strings.readLess : strings.readMore)
                    .font(.system(size: 12, weight: .bold))
                    .foregroundColor(.meTontRed)
                    .padding(.top, 4)
                    .onTapGesture { expanded.toggle() }

                if let websiteUrl = event.websiteUrl, !websiteUrl.isEmpty {
                    Spacer().frame(height: 8)
                    Button(action: {
                        var urlString = websiteUrl
                        if !urlString.hasPrefix("http://") && !urlString.hasPrefix("https://") {
                            urlString = "https://\(urlString)"
                        }
                        if let url = URL(string: urlString) {
                            UIApplication.shared.open(url)
                        }
                    }) {
                        HStack(spacing: 4) {
                            Image(systemName: "globe").font(.system(size: 12))
                            Text(strings.viewEventWebsite).font(.system(size: 12, weight: .medium))
                        }
                        .foregroundColor(.meTontRed)
                        .padding(.horizontal, 12)
                        .padding(.vertical, 6)
                        .overlay(RoundedRectangle(cornerRadius: 8).stroke(Color.meTontRed, lineWidth: 1))
                    }
                }
            }
            .padding(16)
        }
        .background(Color.white)
        .cornerRadius(16)
        .shadow(color: Color.black.opacity(0.06), radius: 4, x: 0, y: 2)
    }
}
